File: chor-extraction/src/main/java/ast/sp/nodes/ProcessBehaviour.java
package ast.sp.nodes;

import ast.sp.interfaces.Behaviour;
import ast.sp.interfaces.SPNode;

import java.util.ArrayList;
import java.util.List;

public class ProcessBehaviour implements SPNode {
    private final String process;
    private final List<ProcedureDefinitionSP> procedures;
    private final Behaviour main;

    private List<String> visitedProcedures;

    public ProcessBehaviour(String process, List<ProcedureDefinitionSP> procedures, Behaviour main) {
        this.process = process;
        this.procedures = procedures;
        this.main = main;
        this.visitedProcedures = new ArrayList<>();
    }

    public String getProcess() {
        return process;
    }

    public List<ProcedureDefinitionSP> getProcedures() {
        return procedures;
    }

    public Behaviour getMain() {
        return main;
    }

    public String toString(){
        StringBuilder builder = new StringBuilder();
        builder.append(process + "{");
        for (ProcedureDefinitionSP procedure: procedures) {
            builder.append(procedure.toString());
        }
        return  builder.append("main {" + main.toString() + "}}").toString();
    }

    public List<String> getVisitedProcedures() {
        return visitedProcedures;
    }

    public void setVisitedProcedures(String proc) {
        visitedProcedures.add(proc);
    }

    public void clearVisitedProcedures() {
        visitedProcedures.clear();
    }
}

File: chor-extraction/src/main/java/ast/sp/nodes/ProcessBehaviour.kt
package ast.sp.nodes

import ast.sp.interfaces.Behaviour
import ast.sp.interfaces.SPNode

class ProcessBehaviour(val procedures: HashMap<String, ProcedureDefinitionSP>, var main: Behaviour) : SPNode {

    override fun toString(): String {
        val builder = StringBuilder()
        builder.append("{")
        procedures.forEach { t, u -> builder.append(t).append("{").append(u.toString()).append("} ")}
        builder.append("main {").append(main.toString()).append("}}")

        return builder.toString()
    }

    fun copy(): ProcessBehaviour{
        //val tempmap = mutableMapOf<String, ProcedureDefinitionSP>()
        //procedures.forEach { str, prdef -> tempmap.put(str,prdef.copy())  }

        val temp = procedures.clone() as HashMap<String, ProcedureDefinitionSP>
        assert(procedures.equals(temp))
        val maincopy = main.copy()
        assert(main.equals(maincopy))
        return ProcessBehaviour(temp, main.copy())
    }

    fun equals(pb: ProcessBehaviour): Boolean{
        if (!main.equals(pb.main)) return false
        for (pr in procedures){
            val pbpr = pb.procedures.get(pr.key)
            if (pbpr == null || !pbpr.behaviour.equals(pr.value)) return false
        }
        return true
    }
}
